core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -182,17 +185,21 @@ public void renameAllTempFiles() throws IOException {
      * Closes the current recovery target and waits up to a certain timeout for resources to be freed.
      * Returns true if resetting the recovery was successful, false if the recovery target is already cancelled / failed or marked as done.
      */
-    boolean resetRecovery() throws InterruptedException, IOException {
+    boolean resetRecovery(TimeValue activityTimeout) throws InterruptedException, IOException, TimeoutException {
         if (finished.compareAndSet(false, true)) {
             try {
-                // yes, this is just a logger call in a try-finally block. The reason for this is that resetRecovery is called from
-                // CancellableThreads and we have to make sure that all references to IndexShard are cleaned up before exiting this method
                 logger.debug("reset of recovery with shard {} and id [{}]", shardId, recoveryId);
             } finally {
                 // release the initial reference. recovery files will be cleaned as soon as ref count goes to zero, potentially now.
                 decRef();
             }
-            closedLatch.await();
+            try {
+                if (closedLatch.await(activityTimeout.millis(), TimeUnit.MILLISECONDS) == false) {
```

Comment:
instead of the passing `activityTimeout` can't we just use the internal cancelableThreads and only execute that latch await under in a cancelable fashion?